AOD tables for forward tracks

rpezzi opened the pull request:
This PR summarizes the discussions around AOD tables for forward tracks held up to now. It aims at a single table for all forward tracks in O2. The new forward track tables are mimicking the style of central barrel tracks tables: 
 * The main table contains physics parameters. 
 * Covariances table (not final - further performance and optimization studies pending)
 * Extra table for tracking and detector specific data. This table brings chi2 and matchchi2 from the MUON table. It is not obvious if we should have a single chi2 column for all track types of a dedicated column for each. 
 
@iarsene @vfeuilla , please notice that I did not bring the `RAtAbsorberEnd` and `PDca` colums since these are very specific to MUON and perhaps these should be evaluated at the analysis level. Please take a look and let me know if is there anything else needed to drop the `muon` tables in favor of the `trackfwd` tables.

Ping @jgrosseo @javierecc @ginnocen @aphecetche @pillot @shahor02

rpezzi committed:
Sketching AOD tables for forward tracks

rpezzi committed:
Fix AOD tables for forward tracks

Add cartesian momenta.
New table `TracksExtraFwd` for tracking and detector-specific data.

jgrosseo commented:
Thanks a lot! We will have a detailed look and comment.
(I have changed the PR to draft while we are discussing)

pillot commented:
Actually RAtAbsorberEnd and PDCA cannot be evaluated from other informations stored in the AOD tables. They must be computed at the reconstruction level. They are both used for the standard track selection in the analysis. PDCA will likely be used only for muon-only track analysis, while RAtAbsorberEnd can be used in the muon-mft track selection as well. So they should probably be moved to the trackfwd table if we drop the muon track table.

rpezzi commented:
> Actually RAtAbsorberEnd and PDCA cannot be evaluated from other informations stored in the AOD tables. They must be computed at the reconstruction level. They are both used for the standard track selection in the analysis. PDCA will likely be used only for muon-only track analysis, while RAtAbsorberEnd can be used in the muon-mft track selection as well. So they should probably be moved to the trackfwd table if we drop the muon track table.

@pillot , at the current version of muons AOD tables, both `RAtAbsorberEnd` and `PDCA` are evaluated dynamically at AOD level. See [here](https://github.com/AliceO2Group/AliceO2/blob/dev/Framework/Core/include/Framework/AnalysisDataModel.h#L340-L358). Once evaluated at reconstruction, we should store it at table `TracksExtraFwd`, along specifics for other track types. 

Notice how track-type specific columns for the central barrel are defined [here](https://github.com/AliceO2Group/AliceO2/blob/dev/Framework/Core/include/Framework/AnalysisDataModel.h#L156-L175).

vfeuilla commented:
@rpezzi, Hi, thanks a lot for implementing the forward tracks.

Actually the value of RAtAbsorberEnd that is currently implemented is just a linear extrapolation from the vertex position that I put in to have something to work with and check the implementation of the muon track cuts in the analysis tasks (Because the value wasn't stored in the AOD tables). The value it returns is incorrect. Ultimately it was to be removed.

The correct value should indeed be the one computed at the reconstruction level, abd should be stored somewhere. Doing it at the analysis would require to re-use the Runge-Kutta extrapolation algorithm backwards...

pillot commented:
> Doing it at the analysis would require to re-use the Runge-Kutta extrapolation algorithm backwards...

Hi @vfeuilla . In fact this would not work either because the track parameters in AOD are corrected for absorber effects, which cannot be "de-corrected" (some info are lost). Same problem for the DCA calculation in PDCA.

iarsene commented:
Hi,
The data tables look fine from my side. Regarding the pdca and the radius, as Philippe explained, we should just store these quantities. Since these are quantities that are used in most analyses (at least it was the case in Run-2), i suggest we put them in the main table (trackfwd). Otherwise, these could also go to the extra table, but i suspect we will always use them, so they could stay in the main table.
@vfeuilla then we would also need to adapt the DQ derived tables to reflect the new structure. We should have a discussion also on whether for now we should also include covariance matrix information.
Cheers,
Ionut

rpezzi committed:
Adding RAtAbsorberEnd and PDca to TracksExtraFwd AOD table

Both columns will be evaluated at tracking. At the moment using 
expression columns for debuging purposes.

rpezzi commented:
Hi @iarsene , @vfeuilla 
There is a new commit adding `RAtAbsorberEnd` and `PDca` to TracksExtraFwd table. I rather keep this on the extra table since this will be used for Run2-like MUON Analysis, but not for MFT standalone, GlobalMuonTracks, upgrade studies. The code has been adapted to use expression column: inline evaluation + new coordinate system. Please double check.

jgrosseo commented on Framework/Core/include/Framework/AnalysisDataModel.h:
```diff
@@ -265,6 +265,122 @@ using TrackExtra = TracksExtra::iterator;
+{
+// TRACKPARFWD TABLE definition
+DECLARE_SOA_INDEX_COLUMN(Collision, collision);
+DECLARE_SOA_COLUMN(TrackType, trackType, uint8_t); // TODO change to TrackTypeEnum when enums are supported
```
Please create an enum ForwardTrackTypes for the types in DataTypes.h
then change the name of the enum in the comment

rpezzi replied:
Done. @aphecetche , @pillot notice alias MCHStandalone for Run2-like MUON Tracks in DataTypes.h.

jgrosseo replied:
Thanks. I understand we will have also MCHStandalone for Run 3 ? Otherwise (if it is just legacy) we should add Run2 in the name (as in the other enum) and put it at the end.

pillot replied:
@jgrosseo yes. It is not just legacy in the sense that also in run3 some analysis will use these tracks and not the ones with MFT, in order for instance to profit from the full MUON acceptance (MFT acceptance is smaller).

jgrosseo commented on Framework/AnalysisSupport/src/AODJAlienReaderHelpers.cxx:
```diff
@@ -107,6 +107,12 @@ std::vector<std::string> getColumnNames(header::DataHeader dh)
+      return columnNamesTrait(typename StoredTracksCovMetadata::table_t::persistent_columns_t{});
+    } else if (description == "TRACKFWD:EXTRA") {
+      return columnNamesTrait(typename TracksExtraFwdMetadata::table_t::persistent_columns_t{});
     }
```
We should discuss if we need the split into the three tables which adds overhead on reading for a table. Will there be analysis which do not use the covariance matrix? The extra only has 4 fields at present, so is it worth to have it separate?

rpezzi replied:
Run2 MUON analysis usually does not uses the covariances matrix. MFTstandalone and GlobalMuonTracks are very certainly to use. As for the extra table, I added three missing columns: 7 fields that I can think of at the moment at the extra table. GlobalMuonTracks may need indexes to the corresponding standalone tracks.

rpezzi replied:
The extra table has been incorporated into the main parameters table. Created new table for MFT Standaloneas discussed last week.

jgrosseo commented on Framework/Core/include/Framework/AnalysisDataModel.h:
```diff
@@ -265,6 +265,122 @@ using TrackExtra = TracksExtra::iterator;
+DECLARE_SOA_COLUMN(Chi2, chi2, float);
+DECLARE_SOA_COLUMN(Chi2Match, chi2Match, float);
+//DECLARE_SOA_COLUMN(PDca, pDca, float);
+//DECLARE_SOA_COLUMN(RAtAbsorberEnd, rAtAbsorberEnd, float);
```
Didn't you want to store it (and not calculate on the fly)?

iarsene replied:
Indeed, this is also my understanding from what @pillot explained.

rpezzi committed:
Adding enum for forward tracks types in AODs

rpezzi committed:
Add Chi2s, matching scores and ranking on forward tracks AOD table

shahor02 commented on Framework/Core/include/Framework/AnalysisDataModel.h:
```diff
@@ -265,6 +265,127 @@ using TrackExtra = TracksExtra::iterator;
+//DECLARE_SOA_COLUMN(PDca, pDca, float);
+//DECLARE_SOA_COLUMN(RAtAbsorberEnd, rAtAbsorberEnd, float);
+// Expression columns for PDca and RAtAbsorberEnd will be replaced by SOA columns
+DECLARE_SOA_EXPRESSION_COLUMN(PDca, pDca, float, (aod::trackfwd::p)*nsqrt((aod::trackfwd::x * aod::trackfwd::x) + (aod::trackfwd::y * aod::trackfwd::y) + (aod::trackfwd::z * aod::trackfwd::z)));
```
Irrespective of PDCA discussion above: why z component is accounted here? For prompt tracks PDCA is a measure of effective mult.scattering kick in transverse direction (dcaX, dcaY), while dcaZ is amplified by ``1/tg(theta)``.
Also, do ``aod::trackfwd::x, y`` have meaning of abs. coordinates at vertex or indeed DCA to it?

pillot replied:
@shahor02 as far as I understood, this calculation was just to get something to play with while developing the code, but indeed this is not the way it is computed in reality. It is in fact impossible to compute it with the info stored in AOD.

In the case of muon tracks (without MFT), aod::trackfwd::x, y is equal to the vertex position the track has been extrapolated to, so it cannot be used to compute the DCA. Instead we use the track extrapolated without Branson correction, which is not stored in the AOD.

shahor02 replied:
@pillot OK, but for MFT tracks this computation makes sense, provided the Z is suppressed (and delta to vertex is used).

jgrosseo replied:
So for MFT tracks, we could reduce what is stored and use the computation but not for those which went through the absorber?

rpezzi replied:
For MFTStandalone and GlobalMuonTracks there is no need to store pDCA, as it can be calculated on the fly. Storing pDCA from reconstruction is necessary only for run2-like MCH tracks, which will be renamed to MUONStandalone in `Framework/DataTypes.h` (due popular vote). 

At the moment we do not foresee a table for pDCA evaluated on-the fly for the MFT, but there are several analysis possibilities considered. I mean that we are learning how to use MFTStandalone tracks and GlobalMuonTracks; the optimal data structures are yet to be unveiled.

shahor02 commented on Framework/Core/include/Framework/AnalysisDataModel.h:
```diff
@@ -265,6 +265,127 @@ using TrackExtra = TracksExtra::iterator;
+DECLARE_SOA_COLUMN(Tgl, tgl, float);
+DECLARE_SOA_COLUMN(Signed1Pt, signed1Pt, float);
+
+DECLARE_SOA_DYNAMIC_COLUMN(Charge, charge, [](float signed1Pt) -> short { return (signed1Pt > 0) ? 1 : -1; });
```
Somewhat off-topic: this way of extracting the charge from ``q/pt``, inherited from the barrel tracks (in turn, coming from Aliroot), does not allow proper treatment of charges other than +/- 1. For the forward tracks this might be not a problem, since there we will not consider (is it true also for MFT?) charge 2 tracks, and for B=0 the ``q/pt`` is anyway 0. 
But for the barrel we will have both charge 2 particles and charge 0 particles with measured momentum (V0s).
For this reason I've added https://github.com/AliceO2Group/AliceO2/blob/f14182b852daaa405170019c0b9a7c96830dec5e/DataFormats/Reconstruction/include/ReconstructionDataFormats/TrackParametrization.h#L231 which also should be propagated to AOD.
Perhaps it is worth to introduce it also for forward tracks.
Another variable which might be worth propagating (for barrel tracks only) is the PID assigned during the tracking: https://github.com/AliceO2Group/AliceO2/blob/f14182b852daaa405170019c0b9a7c96830dec5e/DataFormats/Reconstruction/include/ReconstructionDataFormats/TrackParametrization.h#L232, which proved to be useful in understanding problems in the data.

jgrosseo replied:
Thanks for your feedback. Let's bring your remarks up briefly next week at the WP14 meeting. Here some feedback from my side:
Charge:
* The V0s go to a different table, so no worry about that
* Would you assign charge=2 tracks clearly at the reconstruction? Otherwise it would be anyway analysis level
PID hypothesis:
* This looks like a debug / QA information to me as you write also. We could consider adding an extra QA table to the data model which we stop filling at some point. I would refrain from adding it to the standard track table.

rpezzi replied:
I think forward tracks should be prepared for charges other than +-1, but I rather let this for a forthcoming revision of the forward track classes. 
I am adding a new commit naming MUONStandalone track, but will keep the pDCA calculation as it is so that any tests performed by @vfeuilla gives identical results following the migration from muon -> trackfwd.

I think this PR is OK to go.

jgrosseo replied:
I think before merging we should conclude on the discussion about pDca etc on the fly versus storing.
If @vfeuilla wants an exact comparison one can always do a calculation in the analysis task. And this should anyway happen locally. No need to have everything in the dev branch. @vfeuilla can just use your branch for the checks.

So the questions remaining are:
* If we can do on the fly for all but MUONStandalone. How do we split the two cases, and is it worth it.
* What is the expected share of those tracks?
* You also mentioned indices that refer between the tracks where are they. Note that the framework does not yet support indices to the same table, you have to declare them as int for now
* Could you add a comment after each stored member to explain the comment (we will also do that for the rest of the data model)?
* I think there are some unanswered questions from Ruben about the `z` here...

rpezzi replied:
About splitting the cases that need storage from those who do not, perhaps we could implement conditional columns, or have a separate table for MUONStandalone's PDca and RAtAbsorberEnd?

Rought estimates of the shares:

* MUONStandalone is the smallest share of forward tracks (about 4 per PbPb event), which needs PDca and storage. 
* MFTStandalone has the highest storage share, order of 1000 per PbPB event - PDca can be evaluated on the fly.
* GlobalMuonTracks share not know yet. May be about 10 times the MUONStandalone share.  - PDca can be evaluated on the fly.

As for the Z component, I do not know why it is accounted here. O copied the columns used by @vfeuilla tests for a smooth transition from the muon->trackfwd table. @iarsene and @vfeuilla: are you OK in dropping the EXPRESSION_COLUMNs for PDca and RAtAbsorberEnd in favor of the SOA_COLUMNs?

rpezzi committed:
Naming MUONStandalone tracktype on DataTypes.h

jgrosseo commented on Framework/Core/include/Framework/DataTypes.h:
```diff
@@ -28,4 +27,13 @@ enum TrackFlagsRun2Enum {
+{
+enum ForwardTrackTypeEnum : uint8_t {
+  GlobalMuonTrack = 0,
+  MUONtandalone,
```
There is a typo in the name...

rpezzi replied:
fixed.  Marking as ready for review.

jgrosseo reviewed (changes requested):
Sorry, I don't agree that we mark this ready. There are quite some unanswered points in the comments.

rpezzi committed:
Fix dummy columns on forward AOD tables, add comments, better naming

rpezzi commented:
> Sorry, I don't agree that we mark this ready. There are quite some unanswered points in the comments.

Summarizing replies along the discussion above:

* For MUONStandalone (run2-like tracks) there is need to store PDca and RAtAbsorberEnd evaluated during reconstruction.
* For GlobalMuon and MFTStandalone we have what is needed to get started with the current PR.
* Regarding the use of Z on the PDca raised by Ruben, this is being removed since this is a dummy calculation that can be performed at analysis level for development purposes. We are dropping on-the-fly PDca calculation at all.

It seems the only issue is how deal with columns that concerns only a few tracktypes. I will follow up by e-mail to find a slot to discuss.

rpezzi committed:
Add nClusters and IDs of matched MFT and MCH tracks to trackfwd tables

rpezzi committed:
Splitting AOD tables for MFTStandalone and other fwdtracks

This PR implements what has been discussed last week, among other small
fixes:
  - Drop storage of MFTStandalone tracks covariances
  - Add forward track-types
  - Add table for unassigned forward tracks
  - Namespace + descriptions renamed for more natural reading
    - trackfwd -> fwdtrack
  - IDs of MFT and MCH matched tracks
    - integers while self indexing within a table is not implemented
  - Bunch crossing IDs

rpezzi commented:
I could not identify any errors with the o2-dataflow check.

jgrosseo commented on Framework/Core/include/Framework/AnalysisDataModel.h:
```diff
@@ -265,6 +265,134 @@ using TrackExtra = TracksExtra::iterator;
+} // namespace fwdtrack
+
+//Table for MFTStandalone tracks
+DECLARE_SOA_TABLE_FULL(MFTTracks, "MFTTracks", "AOD", "FWDTRACK:MFTPAR",
```
4th parameter: "MFTTRACK"

jgrosseo commented on Framework/AnalysisSupport/src/AODJAlienReaderHelpers.cxx:
```diff
@@ -107,6 +107,12 @@ std::vector<std::string> getColumnNames(header::DataHeader dh)
+      return columnNamesTrait(typename StoredFwdTracksMetadata::table_t::persistent_columns_t{});
+    } else if (description == "FWDTRACK:PARCOV") {
+      return columnNamesTrait(typename StoredFwdTracksCovMetadata::table_t::persistent_columns_t{});
+    } else if (description == "FWDTRACK:MFTPAR") {
```
drop these 2 lines (it is a separate table)

jgrosseo commented on Framework/Core/include/Framework/AnalysisDataModel.h:
```diff
@@ -265,6 +265,134 @@ using TrackExtra = TracksExtra::iterator;
+
+//Table for MFTStandalone tracks
+DECLARE_SOA_TABLE_FULL(MFTTracks, "MFTTracks", "AOD", "FWDTRACK:MFTPAR",
+                       o2::soa::Index<>, fwdtrack::CollisionId, fwdtrack::BCId, fwdtrack::TrackType,
```
why does mftstandalone need the tracktype?

jgrosseo replied:
The TrackType seems still there?!

rpezzi replied:
MFTTracks table does not need a tracktype. Removing.

jgrosseo commented on Framework/Core/include/Framework/DataTypes.h:
```diff
@@ -28,4 +27,16 @@ enum TrackFlagsRun2Enum {
+enum ForwardTrackTypeEnum : uint8_t {
+  MUONStandalone = 0,
+  MFTStandalone,
+  GlobalMuonTrack,
```
GlobalMuonTrackOtherMatch

jgrosseo replied:
You removed the "BestMatch" from the GlobalMuonTrack on purpose?

jgrosseo replied:
Even if a detail, I would start the enum with the most used one, which is GlobalMuonTrack(BestMatch), right?

rpezzi replied:
Moving GlobalMuonTrack to start

jgrosseo commented on Framework/Core/include/Framework/DataTypes.h:
```diff
@@ -28,4 +27,16 @@ enum TrackFlagsRun2Enum {
+{
+enum ForwardTrackTypeEnum : uint8_t {
+  MUONStandalone = 0,
+  MFTStandalone,
```
Not needed

rpezzi committed:
Fixes for mft and fwdtracks AOD tables

  - Drop MFTStandalone from ForwardTrackTypeEnum
  - Adopt GlobalMuonTrackOtherMatch
  - Fix MFTTracks table description

rpezzi commented:
This last commit should address @jgrosseo remarks.

jgrosseo commented on Framework/Core/include/Framework/AnalysisDataModel.h:
```diff
@@ -265,6 +265,134 @@ using TrackExtra = TracksExtra::iterator;
+
+//Table for MFTStandalone tracks
+DECLARE_SOA_TABLE_FULL(MFTTracks, "MFTTracks", "AOD", "MFTTRACK",
+                       o2::soa::Index<>, fwdtrack::CollisionId, fwdtrack::BCId, fwdtrack::TrackType,
```
I think I overlooked this before, but why we have a BCId in MFT? It just needs the CollisionId

rpezzi replied:
Yes, removing BCId from MFTTracks table

jgrosseo commented on Framework/Core/include/Framework/AnalysisDataModel.h:
```diff
@@ -276,6 +404,17 @@ DECLARE_SOA_TABLE(UnassignedTracks, "AOD", "UNASSIGNEDTRACK",
+namespace unassignedfwdtracks
+{
+DECLARE_SOA_INDEX_COLUMN(Collision, collision);
+DECLARE_SOA_INDEX_COLUMN(FwdTrack, fwdtrack);
```
I think this is for MFTTracks and should point to MFTTrack
For the FwdTrack, we can solve the unassigned ones through the BC Index, so we don't need a table for them

rpezzi replied:
I think we need unassigned tracks for both FwdTracks and MFTTracks. FwdTracks will be assigned to BC index when there is MID match. Which is not true for `MCHStandalone` and `GlobalForwardTrack`.

jgrosseo replied:
GlobalForwardTrack has a link to MFT and through that is assigned (or in the unassigned table)
For MCHStandalone @pillot said you would estimate the most likely BC even if there is some uncertainty if I remember the discussion well...

rpezzi replied:
You are right.

rpezzi commented:
Hi @jgrosseo . 
Regarding BC and Collision on MFTTracks, I do not have a clear picture on how this association will happen. Same for unassigned tracks. This will come as we integrate full AOD production for forward tracks. Regarding naming GlobalMuonTracks, I think bestmatch can be omited as this is the main analysis object. Special analysis based on Global Muon tracks other matches will know where to find them. For the rest, at this point I think it will be simpler for you to edit it to your liking.

rpezzi committed:
Further fixes for mft and fwdtracks AOD tables

  - Drop BCId and TrackType from MFTStandalone table  
  - Unnassigned tracks table for MFTTracks instead of FwdTracks
  - SOA_EXTENDED_TABLE for MFTTracks (expression columns)
  - new order on ForwardTrackTypeEnum

jgrosseo commented:
@rpezzi I have imported this PR here: https://github.com/jgrosseo/AliceO2/commit/460333c7a54f1ec8592f5fad94b575f81d8e4d6b
to be able to apply it after the changes in https://github.com/AliceO2Group/AliceO2/pull/5566
Please check if everything is OK. 
Once https://github.com/AliceO2Group/AliceO2/pull/5566 is merged we open a new one with the MFT tables. 
The next step is to adjust the AliPhysics RUN3 converter and remove the old muon tables there, and fill the new ones.

jgrosseo commented:
Closing the discussion here. Thanks for input to everyone.
